iroh-relay/src/key_cache.rs

```diff
@@ -0,0 +1,65 @@
+use std::{
+    num::NonZeroUsize,
+    sync::{Arc, Mutex},
+};
+
+use iroh_base::PublicKey;
+
+type SignatureError = <PublicKey as TryFrom<&'static [u8]>>::Error;
+type PublicKeyBytes = [u8; PublicKey::LENGTH];
+
+/// A cache for public keys.
+///
+/// This is used solely to make parsing public keys from byte slices more
+/// efficient for the very common case where a large number of identical keys
+/// are being parsed, like in the relay server.
+///
+/// The cache stores only successful parse results.
+#[derive(Debug, Clone, Default)]
+pub enum KeyCache {
+    /// The key cache is disabled.
+    #[default]
+    Disabled,
+    /// The key cache is enabled with a fixed capacity. It is shared between
+    /// multiple threads.
+    Shared(Arc<Mutex<lru::LruCache<PublicKey, ()>>>),
+}
+
+impl KeyCache {
+    /// Key cache to be used in tests.
+    #[cfg(test)]
+    pub fn test() -> Self {
```

Comment:
this is the same as `KeyCache::default()`, why is this needed?

Reply:
I was looking through the code for places where KeyCache::default is used to make sure I got them all. So I renamed this to test so it is clear that it is only for tests.

Reply:
I have now done the same for DerpCodec and removed the Default instance. I think with Default it is too easy to accidentally have a place where the cache is not used at all.